File: src/120.ts
import _ from 'lodash';

// Computes the sum of the maximum remainders of (a - 1)^n + (a + 1)^n for 3 ≤ a ≤ 1000.
//
// See {@link https://projecteuler.net/problem=120}.
export default function compute(): number {
  return _.range(3, 1001)
    .map(f)
    .reduce((a, b) => a + b, 0);
}

function f(a: number) {
  function g(n: number) {
    if (n === 1) {
      return 2 * a;
    }

    if (n % 2 === 0) {
      return 2;
    }

    return 2 * n * a;
  }

  return _.range(1, 2 * a + 1)
    .map(g)
    .map(n => n % (a * a))
    .reduce((a, b) => Math.max(a, b), 0);
}

File: project_euler/120.py
# Problem 120: Square Remainders
#
# See https://projecteuler.net/problem=120
def r_max(a: int) -> int:
    return 2 * a * ((a - 1) // 2)


def run(limit=1_000) -> int:
    return sum(r_max(a) for a in range(3, limit + 1))


def test_run():
    assert run() == 333082500
